File: 백준/Silver/2824. 최대공약수/최대공약수.py
# 최대공약수
import sys
input = sys.stdin.readline

def gcd(a, b):
    if a < b:
        a, b = b, a

    while b > 0:
        a, b = b, a % b

    return a

N = int(input())
n_nums = list(map(int,input().split()))

M = int(input())
m_nums = list(map(int,input().split()))

result = 1

for i in range(N):
    for j in range(M):
        gcd_nums = gcd(n_nums[i], m_nums[j])
        result *= gcd_nums
        n_nums[i] //= gcd_nums
        m_nums[j] //= gcd_nums

if len(str(result)) > 9:
    print(str(result)[-9:])
else:
    print(result)

File: 백준/Silver/2824. 최대공약수/최대공약수.java
import java.io.*;
import java.math.BigInteger;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int N = Integer.parseInt(br.readLine());
        BigInteger[] nNums = new BigInteger[N];

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            nNums[i] = new BigInteger(st.nextToken());
        }

        int M = Integer.parseInt(br.readLine());
        BigInteger[] mNums = new BigInteger[M];

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < M; i++) {
            mNums[i] = new BigInteger(st.nextToken());
        }

        BigInteger result = new BigInteger("1");

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                BigInteger gcdNums = gcd(nNums[i], mNums[j]);
                result = result.multiply(gcdNums);
                nNums[i] = nNums[i].divide(gcdNums);
                mNums[j] = mNums[j].divide(gcdNums);
            }
        }

        String stringResult = result.toString();

        if (stringResult.length() > 9) {
            System.out.println(stringResult.substring(stringResult.length()-9));
        } else {
            System.out.println(stringResult);
        }


    }

    static BigInteger gcd(BigInteger a, BigInteger b) {
        if (b.compareTo(BigInteger.ZERO) <= 0) {
            return a;
        }

        return gcd(b, a.mod(b));
    }
}
